core/src/main/java/io/questdb/cutlass/line/tcp/LineTcpMeasurementScheduler.java

```diff
@@ -0,0 +1,812 @@
+                    highestLoadedThreadId = n;
+                }
+
+                if (lowestLoad > loadByThread[n]) {
+                    lowestLoad = loadByThread[n];
+                    lowestLoadedThreadId = n;
+                }
+            }
+
+            if (highestLoadedThreadId == -1 || lowestLoadedThreadId == -1 || highestLoadedThreadId == lowestLoadedThreadId) {
+                break;
+            }
+
+            double loadRatio = (double) highestLoad / (double) lowestLoad;
+            if (loadRatio < maxLoadRatio) {
+                // Load is not sufficiently unbalanced
+                break;
+            }
+
+            int nTables = 0;
+            lowestLoad = Integer.MAX_VALUE;
+            String leastLoadedTableName = null;
+            for (int n = 0, sz = tableNames.size(); n < sz; n++) {
+                TableUpdateDetails stats = tableUpdateDetailsByTableName.get(tableNames.get(n));
+                if (stats.threadId == highestLoadedThreadId && stats.nUpdates > 0) {
+                    nTables++;
+                    if (stats.nUpdates < lowestLoad) {
+                        lowestLoad = stats.nUpdates;
+                        leastLoadedTableName = stats.tableName;
+                    }
+                }
+            }
+
+            if (nTables < 2) {
+                // The most loaded thread only has 1 table with load assigned to it
+                maxLoad = highestLoad;
+                continue;
+            }
+
+            fromThreadId = highestLoadedThreadId;
+            toThreadId = lowestLoadedThreadId;
+            tableNameToMove = leastLoadedTableName;
+            break;
+        }
+
+        for (int n = 0, sz = tableNames.size(); n < sz; n++) {
+            TableUpdateDetails stats = tableUpdateDetailsByTableName.get(tableNames.get(n));
+            stats.nUpdates = 0;
+        }
+
+        if (null != tableNameToMove) {
+            LineTcpMeasurementEvent event = getNewEvent();
+            if (null == event) {
+                return;
+            }
+            LOG.info().$("rebalance cycle ").$(++nRebalances).$(" moving ").$(tableNameToMove).$(" from ").$(fromThreadId).$(" to ").$(toThreadId).$();
+            commitRebalanceEvent(event, fromThreadId, toThreadId, tableNameToMove);
+            TableUpdateDetails stats = tableUpdateDetailsByTableName.get(tableNameToMove);
+            stats.threadId = toThreadId;
+        }
+    }
+
+    private void calcThreadLoad() {
+        Arrays.fill(loadByThread, 0);
+        ObjList<CharSequence> tableNames = tableUpdateDetailsByTableName.keys();
+        for (int n = 0, sz = tableNames.size(); n < sz; n++) {
+            TableUpdateDetails stats = tableUpdateDetailsByTableName.get(tableNames.get(n));
+            loadByThread[stats.threadId] += stats.nUpdates;
+        }
+    }
+
+    int[] getLoadByThread() {
+        return loadByThread;
+    }
+
+    int getnRebalances() {
+        return nRebalances;
```

Comment:
`rebalances`  is not a word. It gets consistently highlighted with spell check. Perhaps we can create a different term?

Reply:
https://dictionary.cambridge.org/dictionary/english/rebalance

Reply:
haha, intellij doesn't agree